src/Interpreters/HashJoin/HashJoinMethodsImpl.h

```diff
@@ -379,24 +381,241 @@ size_t HashJoinMethods<KIND, STRICTNESS, MapsTemplate>::joinRightColumnsSwitchMu
+            if constexpr (flag_per_row)
+                used_flags.template setUsed<join_features.need_flags, flag_per_row>(row_ref->columns, row_ref->row_num, 0);
+            else
+                used_flags.template setUsed<join_features.need_flags, flag_per_row>(find_result);
+
+            added_columns.appendFromBlock(row_ref, join_features.add_missing);
+        }
+        else
+            addNotFoundRow<join_features.add_missing, join_features.need_replication>(added_columns, current_offset);
+    }
+    else if constexpr (join_features.is_all_join)
+    {
+        setUsed<need_filter>(added_columns.filter, i);
+        used_flags.template setUsed<join_features.need_flags, flag_per_row>(find_result);
+        auto used_flags_opt = join_features.need_flags ? &used_flags : nullptr;
+        addFoundRowAll<Map, join_features.add_missing>(mapped, added_columns, current_offset, known_rows, used_flags_opt);
+    }
+    else if constexpr ((join_features.is_any_join || join_features.is_semi_join) && join_features.right)
+    {
+        /// Use first appeared left key + it needs left columns replication
+        bool used_once = used_flags.template setUsedOnce<join_features.need_flags, flag_per_row>(find_result);
+        if (used_once)
+        {
+            auto used_flags_opt = join_features.need_flags ? &used_flags : nullptr;
+            setUsed<need_filter>(added_columns.filter, i);
+            addFoundRowAll<Map, join_features.add_missing>(mapped, added_columns, current_offset, known_rows, used_flags_opt);
+        }
+    }
+    else if constexpr (join_features.is_any_join && join_features.inner)
+    {
+        bool used_once = used_flags.template setUsedOnce<join_features.need_flags, flag_per_row>(find_result);
+
+        /// Use first appeared left key only
+        if (used_once)
+        {
+            setUsed<need_filter>(added_columns.filter, i);
+            added_columns.appendFromBlock(&mapped, join_features.add_missing);
+        }
+    }
+    else if constexpr (join_features.is_any_join && join_features.full)
+    {
+        /// TODO
+    }
+    else if constexpr (join_features.is_anti_join)
+    {
+        if constexpr (join_features.right && join_features.need_flags)
+            used_flags.template setUsed<join_features.need_flags, flag_per_row>(find_result);
+    }
+    else /// ANY LEFT, SEMI LEFT, old ANY (RightAny)
+    {
+        setUsed<need_filter>(added_columns.filter, i);
+        used_flags.template setUsed<join_features.need_flags, flag_per_row>(find_result);
+        added_columns.appendFromBlock(&mapped, join_features.add_missing);
     }
 }
 
+template <JoinKind KIND, JoinStrictness STRICTNESS, typename MapsTemplate>
+template <
+    typename KeyGetter,
+    typename Map,
+    bool need_filter,
+    bool check_null_map,
+    JoinCommon::JoinMask::Kind join_mask_kind,
+    typename AddedColumns,
+    typename Selector>
+size_t HashJoinMethods<KIND, STRICTNESS, MapsTemplate>::joinRightColumns(
```

Comment:
It has the same code as the generic `joinRightColumns`, but specialised for the case of one condition (i.e. one map)